bot/glossary_command.go

```diff
@@ -0,0 +1,149 @@
+package bot
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/quintilesims/slackbot/db"
+	"github.com/quintilesims/slackbot/models"
+	glob "github.com/ryanuber/go-glob"
+	"github.com/urfave/cli"
+)
+
+// NewGlossaryCommand returns a cli.Command that manages !glossary
+func NewGlossaryCommand(store db.Store, w io.Writer) cli.Command {
+	return cli.Command{
+		Name:  "!glossary",
+		Usage: "manage the glossary",
+		Subcommands: []cli.Command{
+			{
+				Name:      "define",
+				Usage:     "add or set an entry in the glossary",
+				ArgsUsage: "KEY DEFINITION",
+				Action:    newGlossaryDefineAction(store, w),
+			},
+			{
+				Name:      "rm",
+				Usage:     "remove an entry from the glossary",
+				ArgsUsage: "KEY",
+				Action:    newGlossaryRemoveAction(store, w),
+			},
+			{
+				Name:      "search",
+				Usage:     "search for entries in the glossary",
+				ArgsUsage: "GLOB",
+				Action:    newGlossarySearchAction(store, w),
+			},
+		},
+	}
+}
+
+func newGlossaryDefineAction(store db.Store, w io.Writer) func(c *cli.Context) error {
+	return func(c *cli.Context) error {
+		args := c.Args()
+		key := args.First()
+		definition := strings.Join(args.Tail(), " ")
+
+		if key == "" {
+			return fmt.Errorf("Arg KEY is required")
+		}
+
+		if definition == "" {
+			return fmt.Errorf("Arg DEFINITION is required")
+		}
+
+		glossary := models.Glossary{}
+		if err := store.Read(db.GlossaryKey, &glossary); err != nil {
+			return err
+		}
+
+		var text string
+		if previousDefinition, ok := glossary[key]; previousDefinition != "" && ok {
+			text = fmt.Sprintf("Replacing previous definition of *%s*: *%s* with *%s*", key, previousDefinition, definition)
+		} else {
+			text = fmt.Sprintf("Ok, *%s* %s", key, definition)
+		}
+
+		glossary[key] = definition
+
+		if err := store.Write(db.GlossaryKey, glossary); err != nil {
+			return err
+		}
+
+		if _, err := w.Write([]byte(text)); err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
+func newGlossaryRemoveAction(store db.Store, w io.Writer) func(c *cli.Context) error {
+	return func(c *cli.Context) error {
+		key := c.Args().Get(0)
+
+		if key == "" {
+			return fmt.Errorf("Arg KEY is required")
+		}
+
+		glossary := models.Glossary{}
+		if err := store.Read(db.GlossaryKey, &glossary); err != nil {
+			return err
+		}
+
+		if _, ok := glossary[key]; ok {
+			delete(glossary, key)
+		} else {
+			return fmt.Errorf("Key: *%s* not in glossary", key)
+		}
+
+		if err := store.Write(db.GlossaryKey, glossary); err != nil {
```

Comment:
What are you thoughts on adding a delete to db? I think it'd be clearer, even if it's just a wrapper around Write.

Reply:
I don't think that would work the way you think it would. With the current way `db.Store` is written, that would only allow deleting entire tables, like clearing out the entire `glossary`, not individual entries within the glossary like you'd want. 

The `db.Store` interface is extremely simple and flexible, and I don't think we should change that interface unless we hit performance concerns or something like that (even then, I have ideas on how to improve that w/o changing the interface).

Reply:
Could you compare the incoming table with the previous one, to see if something is removed or not. It's low priority, I just think it'd read better.